internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,470 @@
+	// or when it expires, or when the key changed).
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if err := m.createDir(); err != nil {
+		return nil, fmt.Errorf("cannot create TLS directory: %w", err)
+	}
+	if err := m.getIDCert(); err != nil {
+		return nil, fmt.Errorf("cannot get identity certificate: %w", err)
+	}
+	if err := m.createTLSCert(); err != nil {
+		return nil, fmt.Errorf("cannot create TLS certificate: %w", err)
+	}
+	return m.tlsCert, nil
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *TLSManager) Ensure() error {
+	return nil
+}
+
+func (m *TLSManager) createTLSCert() error {
+	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return err
+	}
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return err
+	}
+	now := timeNow()
+	template := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               defaultCertSubject(m.signer.Fingerprint()),
+		NotBefore:             now,
+		NotAfter:              now.Add(tlsCertValidity),
+		KeyUsage:              x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+
+	// If an externally supplied TLS certificate template was provided,
+	// use it to update the supported fields.
+	if m.tlsTemplate != nil {
+		template.Subject = deepCopyName(m.tlsTemplate.Subject)
+		template.Issuer = deepCopyName(m.tlsTemplate.Issuer)
+		// Supported SAN (Subject Alternate Name) fields.
+		template.DNSNames = slices.Clone(m.tlsTemplate.DNSNames)
+		template.EmailAddresses = slices.Clone(m.tlsTemplate.EmailAddresses)
+	}
+
+	// DER encoded bytes
+	derBytes, err := x509.CreateCertificate(rand.Reader, &template, m.idCert, privateKey.Public(), m.signer)
+	if err != nil {
+		return err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	certificate, err := x509.ParseCertificate(derBytes)
+	if err != nil {
+		return err
+	}
+	m.tlsCert = &tls.Certificate{
+		PrivateKey: privateKey,
+		// Leaf first, then CA cert.
+		Certificate: [][]byte{certificate.Raw, m.idCert.Raw},
+		Leaf:        certificate,
+	}
+	return nil
+}
+
+// createDir verifies the directory layout and permissions, and creates
+// the leaf element of the TLS directory if it does not yet exist.
+func (m *TLSManager) createDir() error {
+	exists, err := pathExists(m.tlsDir)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return expectPermission(m.tlsDir, 0o700)
+	}
+
+	// Create the leaf directory node with 0o700 permissions.
+	return os.Mkdir(m.tlsDir, 0o700)
+}
+
+// getIDCert verifies that the identity certificate exists, and that the
+// certificate matches the supplied identity key signer. If the identity key
+// has changed, or the certificate has expired, the identity certificate is
+// replaced with a new identity certificate signed with the identity key.
+func (m *TLSManager) getIDCert() error {
```

Comment:
This does not return a certificate, it should have a better name.
```suggestion
func (m *TLSManager) ensureIDCert() error {
```

Reply:
Fixed, thank you. Resolving.